### PR TITLE
Stop using bindgen and instead provide a platform-independent static …

### PR DESCRIPTION
**Pull request by sdroege**
…version of the FFI definitions

bindgen requires libclang to be available and on various platforms providing a compatible version of libclang is not that easy and unnecessarily complicates the build.

The dav1d API is relatively stable nowadays so manual updates to the FFI definitions shouldn't cause too much additional work here.

----

Context of this change is https://gitlab.freedesktop.org/gstreamer/cerbero/-/merge_requests/954

### CODE REVIEW

File: dav1d-sys/src/lib.rs
Comment (sdroege) [inline dav1d-sys/src/lib.rs:76]: This is all still based on the bindgen struct definitions and I would clean this up a bit more with regards to the naming

File: dav1d-sys/src/lib.rs
Comment (sdroege) [inline dav1d-sys/src/lib.rs:543]: This should probably just be a `#[repr(transparent)]` struct around `c_void`

File: dav1d-sys/src/lib.rs
Comment (sdroege) [inline dav1d-sys/src/lib.rs:592]: This should probably just be a `#[repr(transparent)]` struct around `c_void`, and especially the way it was generated by bindgen here is actually wrong

File: dav1d-sys/src/lib.rs
Comment (sdroege) [inline dav1d-sys/src/lib.rs:387]: Will add `use std::os::raw::c_int` etc at the top so there's less line-noise everywhere

File: dav1d-sys/src/lib.rs
Comment (sdroege) [inline dav1d-sys/src/lib.rs:23]: Stylistic question, but IMHO changing all the enum constants in the following way would reduce noise (they're double-prefixed currently)

```suggestion
pub const Dav1dObuType_SEQ_HDR: Dav1dObuType = 1;
```

File: dav1d-sys/src/lib.rs
Comment (sdroege) [inline dav1d-sys/src/lib.rs:496]: Still need to double-check all the struct layouts manually against the headers

**Comment by sdroege** [comment]
> The dav1d API is relatively stable nowadays so manual updates to the FFI definitions shouldn't cause too much additional work here.

Basically whenever a new dav1d release is made, you would diff the headers and then add/change whatever has changed. Which should be manageable.

It would also allow you to make the APIs available conditionally based on the dav1d version it is actually compiled with, providing more control to users about the dav1d version they're targetting (compare: gtk-rs version feature flags that allow a user to opt-in explicitly to having e.g glib 2.72 APIs available)

File: Cargo.toml
Comment (lu-zero) [inline Cargo.toml:14]: Why is it needed?

**Comment by lu-zero** [comment]
I like the idea, and I thank you a lot for the effort since it will take quite a bit of time.

Some random notes
- Ideally https://doc.rust-lang.org/core/ffi/index.html has the types we should use, not sure if it is better to wait and see if a newer bindgen will do that for us or it is better to do a replacement pass on the current code.
- I'm not totally convinced we want to remove the ability to use bindgen to freshen the binding, if core::ffi is enough to take care of the platform-specific details, probably it would be fine anyway though.

File: Cargo.toml
Comment (lu-zero) [reply Cargo.toml:14]: maybe it is nicer if you put it as dependency for dav1d and re-export from there.

File: src/lib.rs
Comment (lu-zero) [reply src/lib.rs:51]: Yes and possibly hide the use of libc error numbers.

File: dav1d-sys/src/lib.rs
Comment (lu-zero) [reply dav1d-sys/src/lib.rs:387]: use `core::ffi:c_int` maybe?

File: src/lib.rs
Comment (sdroege) [reply src/lib.rs:51]: You're thinking of having specific constants for those 5 or so errnos that dav1d is using, e.g. the following in dav1d-sys:

```rust
pub const DAV1D_ERR_EAGAIN: c_int = if libc::EPERM < 0 { -libc::EAGAIN } else { libc::EAGAIN };
```
?

File: Cargo.toml
Comment (sdroege) [reply Cargo.toml:14]: I agree, something like https://github.com/rust-av/dav1d-rs/pull/68#discussion_r988808325

File: dav1d-sys/src/lib.rs
Comment (sdroege) [reply dav1d-sys/src/lib.rs:387]: I used `std::os::raw` here because otherwise we'd depend on Rust >= 1.64 just for that. If that's OK for you I can change it accordingly.

**Comment by sdroege** [comment]
> Ideally https://doc.rust-lang.org/core/ffi/index.html has the types we should use, not sure if it is better to wait and see if a newer bindgen will do that for us or it is better to do a replacement pass on the current code.

That's just a matter of search&replace, doesn't really matter what bindgen does here. I can change it to use those types if you prefer but see https://github.com/rust-av/dav1d-rs/pull/68#discussion_r988809966 (we'd have to depend on Rust >= 1.64 just for that and it's not like dav1d-the-C-library will work in no-std environments).

> I'm not totally convinced we want to remove the ability to use bindgen to freshen the binding, if core::ffi is enough to take care of the platform-specific details, probably it would be fine anyway though.

The only platform-specific details in dav1d are the errnos, which are statically covered by the libc crate already (which is maintained by the Rust team). The dav1d C ABI is fortunately simple.

The additional complications that bindgen adds here doesn't really seem worth it, and by not using it we get the ability to actually give API-stability guarantees for dav1d-sys which is IMHO a huge improvement.

**Comment by sdroege** [comment]
Before continuing here, I'd need a decision from you whether to keep bindgen in some form (and not give any API guarantees about dav1d-sys) or if you're OK with dropping it. FWIW, I'm fine making sure that dav1d-sys stays in sync with the dav1d C API in the future so that's no additional work for you at least.

File: src/lib.rs
Comment (lu-zero) [reply src/lib.rs:51]: Yes :)

**Comment by lu-zero** [comment]
I think we can drop it, hide the libc dependency in the `-sys` crate and document in the readme that bindgen is not used.

**Comment by sdroege** [comment]
Sounds good, I'll update this accordingly later. Thanks!

**Comment by sdroege** [comment]
> document in the readme that bindgen is not used

How/why would you document that btw? We're also not using many other things :)

File: dav1d-sys/src/lib.rs
Comment (sdroege) [reply dav1d-sys/src/lib.rs:23]: Actually 

```suggestion
pub const DAV1D_OBU_SEQ_HDR: Dav1dObuType = 1;
```

That way we can even get rid of some style warnings that are silenced currently.

**Comment by sdroege** [comment]
Updated that all in easily to review separate commits. From my side this would be ready, please let me know what you would like to have changed :)

**Review by lu-zero — APPROVED** [review]
It looks neat!

**Comment by sdroege** [comment]
Should I squash all commits into one or do you want to merge it like this?

Also it would be great if we could make a release after this to unblock GStreamer shipping this :) dav1d could become 0.9.1 (no API changes) and dav1d-sys 0.7.0 (API changes).

**Comment by lu-zero** [comment]
Up to you, and yes once it lands I'll cut a release.

**Comment by sdroege** [comment]
I prefer like this because it makes the individual changes easier to follow for anybody who wants to follow what changed

**Pull request merged**
